src/extra/utils/ftd.pyx

```diff
@@ -0,0 +1,517 @@
+                    break
+
+    return np.asarray(edges)[:edge_idx], np.asarray(amplitudes)[:edge_idx], \
+        edge_idx
+
+
+def dled(
+    data_array_t signal, data_t threshold, data_t ratio_max=0.6, data_t width=0,
+    int interp = EdgeInterpolation.LINEAR,
+    data_array_t edges = None, data_array_t amplitudes = None
+):
+    """Dynamic leading edge discriminator.
+
+    This algorithm finds the leading edge of pulses in a 1D signal
+    located at a certain ratio to the pulse's peak value.
+
+    As with the constant fraction discriminator, it is assumed the
+    rising pulse slope points away from zero.
+
+    Args:
+        signal (ArrayLike): 1D input array with analog signal.
+        threshold (data-type): Trigger threshold, positive values imply
+            a positive pulse slope while negative values correspondingly
+            imply a negative pulse slope.
+        ratio_max (data-type, optional): Ratio of leading edge to peak
+            value, 0.6 by default.
+        width (data-type, optional): Minimal distance between found
+            edges, none by default.
+        interp (int, optional): Interpolation mode to locate the edge
+            position, linear by default.
+        edges (ArrayLike, optional): 1D output array to hold the
+            positions of found edges, a new one is allocated if None is
+            passed.
+        amplitudes (ArrayLike, optional): 1D output array to hold the
+            pulse amplitudes corresponding to found edges, a new one is
+            allocated if None is passed.
+    Returns:
+        (ArrayLike, ArrayLike, int) 1D arrays containing the edge
+            positions and amplitudes, number of found edges.
+    """
+
+    if edges is None:
+        edges = np.zeros(
+            len(amplitudes) if amplitudes is not None else len(signal) // 100,
+            dtype=np.asarray(signal).dtype)
+
+    if amplitudes is None:
+        amplitudes = np.zeros_like(edges, dtype=np.asarray(signal).dtype)
+
+    # Negative thresholds require inverted inequality relations when
+    # comparing those to signals. For optimal performance without
+    # duplicating most of the function body, the threshold and signal
+    # are inverted instead (essentially making the signal always
+    # positive) or handled specifically in other cases.
+    cdef bint negative = threshold < 0
+    cdef data_t s = 1
+
+    if negative:
+        threshold = -threshold
+        s = -1
+
+    cdef int signal_idx, signal_len = signal.shape[0], \
+        edge_idx = 0, max_edge = min(edges.shape[0], edges.shape[0]), \
+        ratio_idx, peak_idx, last_peak_idx = 0
+
+    cdef data_t cur_value, ratio_pos, ratio_value, peak_value = 0.0, \
+        last_ratio_pos = -width
+    cdef bint beyond_threshold = False
+
+    with nogil:
+        for signal_idx in range(signal_len):
+            cur_value = signal[signal_idx]
+
+            if negative:
+                cur_value = -cur_value
```

Comment:
```suggestion
            cur_value = s * signal[signal_idx]
```

I would think this is quicker than the extra `if`?

Reply:
Hmmm, not sure if I did profile it or not. Can give it another try.

Reply:
You're right, in particular for negative threholds (which are more common).